Routine, Task 접근 시 권한 체크 기능 추가

ybell1028 opened the pull request:
## Routine, Task 접근 시 권한을 체크하는 Interceptor 등록

HttpMethod가
- `GET`이고 `pathVariables`이 비어있을때 == id로 단일 조회할때
- `PUT`일때, `DELETE`일때

UserId 비교해서 같으면 pass, 틀리면 Exception 던집니다.

제네릭이랑 인터페이스 잘쓰면 클래스 하나로 관리할 수도 있을 것 같았는데 

```Parameter 0 of constructor in org.imgoing.api.application.interceptor.PermissionInterceptor required a single bean, but 2 were found:```

에러가 뜨면서 DI에 필요한 Bean을 찾지 못하는 이슈가 있었음.
같은 타입의 Bean이 컨테이너에 여러개 있어서 저런 에러가 뜨는 것으로 사료됨.
@Primary나 @Qualifier를 써보라는 답변이 많은데 전자는 하나밖에 지정을 못하고 후자는 따로 나눠놓는 거랑 다를게 없으니, 그냥 분리시켜놨음.

리팩토링 가능하다면 부탁드립니다!


## 인터셉터를 등록하면 추가적으로 한번씩 더 조회하게 되는데, 성능 문제는 없을까?

- 딱히 쿼리를 안날려도 되는 삭제 API에는 Builder로 인스턴스를 생성해서 삭제하는 방식으로 변경 (이미 인터셉터에서 해당 데이터가 존재하는 것과 소유자라는 것이 증명되었으니)

### 그럼 나머지의 경우는?
- 수정시 삭제와 비슷하게 처리하면 Auditing이 null이 되버리는 이슈 존재
- Routine과 Task(정확히는 Routinetask)는 언제나 같이 다니는 사이라 즉시로딩으로 설정하고 참조하는 일이 잦음 -> Service에서 조회해오지 않으면 불편 & 비효율적
-  개발 편의성과 성능의 Tradeoff로 남는가, 개선 방법이 없을까?

## etc.

- Mapper 인터페이스 명 변경

ybell1028 committed:
Refactor : Routine, Task 수정 삭제 리팩토링, Mapper 인터페이스 명 변경

- RoutinetaskService.update를 RoutineService 내부로 이동
- 삭제시 데이터 조회하지 않고 인스턴스 build하여 삭제

ybell1028 committed:
Refactor : Routinetask 정렬 코드 수정, compareTo 메소드 삭제

ybell1028 committed:
Feat : Routine, Task 접근 시 권한 체크 기능 추가

ujin2021 commented on src/main/java/org/imgoing/api/application/interceptor/RoutinePermissionInterceptor.java:
```diff
@@ -0,0 +1,46 @@
+            Long id = Long.parseLong((String)pathVariables.get(entityName + "Id"));
+
+            Routine data = routineService.getById(id);
+            if(data.getUser().getId() == payload.getId()) {
```
routineRepository의 findById를 바로 사용안하고 service를 사용하는 이유가 있어?? 해당 루틴이 없을 때도 처리해주려고 그런거야??

ybell1028 replied:
나도 굳이 Service를 주입할 필요가 있을까 하는 고민을 했었어
오늘 질문해줬던 부분에 대해서 정리한 것들을 생각해보면 **Interceptor는 Controller를 호출하기 전과 후에 요청과 응답을 참조하거나 가공할 수 있는 기능**인데, preHandle은 전자에 해당하지.
즉, Controller에 도달하기 전 처리를 해주는 메소드인데 Service를 사용하는건 부적절해보이네. Repository를 사용하는걸로 변경하는게 좋겠다!

ybell1028 committed:
Refactor : 접근 권한 인터셉터 의존성 수정

ujin2021 reviewed (approved):
😊👍

bluayer commented on src/main/java/org/imgoing/api/application/interceptor/RoutinePermissionInterceptor.java:
```diff
@@ -0,0 +1,46 @@
+
+            Long id = Long.parseLong((String)pathVariables.get("routineId"));
+
+            Routine data = routineRepository.findById(id)
```
거의 상관은 없지만 명명이 data인게 좀 아쉽ㅋㅋㅋ

bluayer commented on src/main/java/org/imgoing/api/application/interceptor/RoutinePermissionInterceptor.java:
```diff
@@ -0,0 +1,46 @@
+        String requestHttpMethod = request.getMethod();
+        Map<?, ?> pathVariables = (Map<?, ?>) request.getAttribute(HandlerMapping.URI_TEMPLATE_VARIABLES_ATTRIBUTE);
+
+        if( (HttpMethod.GET.matches(requestHttpMethod) && !pathVariables.isEmpty())
```
if( 옆에 띄어쓰기 의도한게 아니라면 지워주세요!

bluayer commented on src/main/java/org/imgoing/api/application/interceptor/RoutinePermissionInterceptor.java:
```diff
@@ -0,0 +1,46 @@
+            if(data.getUser().getId() == payload.getId()) {
+                return true;
+            }
+            throw new ImgoingException(ImgoingError.BAD_REQUEST, "접근할 수 없는 컨텐츠 입니다.");
```
중첩된 if else 문 안에서 if문에서 return하는 경우에는 else 문을 넣어줘서 반대급부로는 어떤 동작을 하는지 명확하게 보여주는게 좋아!
밑의 라인에서 else 준 것처럼!

bluayer commented on src/main/java/org/imgoing/api/application/interceptor/TaskPermissionInterceptor.java:
```diff
@@ -0,0 +1,46 @@
+            Task data = taskRepository.findById(id)
+                    .orElseThrow(() -> new ImgoingException(ImgoingError.BAD_REQUEST, "존재하지 않는 준비항목입니다."));
+
+            if(data.getUser().getId() == payload.getId()) {
+                return true;
+            }
+            throw new ImgoingException(ImgoingError.BAD_REQUEST, "접근할 수 없는 컨텐츠 입니다.");
+        } else return true;
```
원라인 일 때 {}를 생략하고 싶은지 아니면 중괄호를 쓰고 싶은지 통일해주자!
(Routine쪽에서도!)

bluayer commented on src/main/java/org/imgoing/api/controller/RoutineController.java:
```diff
@@ -69,8 +65,8 @@
     ) {
-        Routine routine = routineService.getById(id);
-        String responseMessage = "루틴 " + routine.getName() + " 이(가) 삭제되었습니다.";
+        Routine routine = Routine.builder().id(id).build();
+        String responseMessage = "루틴이 삭제되었습니다.";
```
prehandle에서 Request에 entity를 담아서 보내는건 어때?
그러면 굳이 이렇게 빌드할 필요가 없을꺼 같은데

bluayer replied:
이거는 Routine, Task 둘 다!

bluayer commented on src/main/java/org/imgoing/api/service/RoutineService.java:
```diff
@@ -41,11 +42,18 @@ public void delete(Routine routine){
+    public Routine update(Routine updateRoutine, List<Long> updateTaskIdList){ // mapper에 의해 새로 생성된 Routine
+        Routine oldRoutine = getById(updateRoutine.getId());
+        oldRoutine.modifyRoutine(updateRoutine);
+        Routine newRoutine = routineRepository.save(oldRoutine);
```
사실 아마 이렇게 명시적으로 저장하지 않고 oldRoutine을 사용해도 될꺼 같은데?
내가 잘못 생각하고 있는걸까?

bluayer commented on src/main/java/org/imgoing/api/domain/entity/Routine.java:
```diff
@@ -38,7 +38,7 @@ public void modifyRoutine(Routine newRoutine) {
 
     public List<Routinetask> getRoutinetasks() {
-        if(routinetasks != null) Collections.sort(routinetasks);
+        if(routinetasks != null) routinetasks.sort(Comparator.comparing(Routinetask::getPriority));
```
routinetasks가 null인 경우가 있나? 애초에 new ArrayList<>()로 생성하는데!
.isEmpty()가 더 적합해보여!

bluayer commented on src/main/java/org/imgoing/api/service/RoutineService.java:
```diff
@@ -41,11 +42,18 @@ public void delete(Routine routine){
+        oldRoutine.modifyRoutine(updateRoutine);
+        Routine newRoutine = routineRepository.save(oldRoutine);
+
+        List<Routinetask> routinetasks = routinetaskService.update(newRoutine, updateTaskIdList);
```
update보다 더 시스템적으로 의미를 가지는 이름이 없을지 한 번만 생각해주기 :)

bluayer reviewed (changes requested):
인증/인가할 때 보통 Interceptor를 쓰는게 확실히 일반적인거 같아!
작업 충분히 잘해줬고, 내가 남긴 코멘트 적용해서 더 나은 방향으로 작성해보자!

ybell1028 committed:
Refactor : PermissionInterceptor 코드 수정

ybell1028 committed:
Refactor : Interceptor에서 조회한 Task HttpServletRequest에 추가하여 사용

ybell1028 committed:
Refactor : Interceptor에서 조회한 Routine HttpServletRequest에 추가하여 사용

- data 변수명 변경
- if-else문 명확하게 표기, 유저 id 확인 부분 중괄호 삭제 원라인으로 변경
- update 로직 변경
- routinetasks null 체크 -> !isEmpty()로 변경

ybell1028 committed:
Refactor : TaskService.update -> modify 이름 변경

ybell1028 committed:
Refactor : RoutineService.update -> modify 이름 변경
